File: frontend/src/api/personalEvents.ts
import { authFetch } from "./client";

export type PersonalEvent = {
  id: string;
  title: string;
  start: string;
  end: string | null;
  all_day: boolean;
  color: string | null;
  location: string | null;
  description: string | null;
};

export async function fetchPersonalEvents(): Promise<PersonalEvent[]> {
  const res = await authFetch("/api/personal-events");
  if (!res.ok) throw new Error("個人予定の取得に失敗しました");
  return res.json();
}

export type PersonalEventInput = {
  title: string;
  start: string;
  end: string | null;
  all_day: boolean;
  color: string | null;
  location?: string | null;
  description?: string | null;
};

export async function createPersonalEvent(event: PersonalEventInput): Promise<PersonalEvent> {
  const res = await authFetch("/api/personal-events", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(event),
  });
  if (!res.ok) throw new Error("個人予定の追加に失敗しました");
  return res.json();
}

export async function updatePersonalEvent(id: string, event: PersonalEventInput): Promise<PersonalEvent> {
  const res = await authFetch(`/api/personal-events/${id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(event),
  });
  if (!res.ok) throw new Error("個人予定の更新に失敗しました");
  return res.json();
}

export async function deletePersonalEvent(id: string): Promise<void> {
  const res = await authFetch(`/api/personal-events/${id}`, {
    method: "DELETE",
  });
  if (!res.ok) throw new Error("個人予定の削除に失敗しました");
}

File: backend/app/api/personal_events.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy.orm import Session

from app.core.auth import CurrentUser, ensure_profile, get_current_user
from app.db.session import get_db
from app.models.personal_event import PersonalEvent
from app.schemas.task import CreatePersonalEvent, PersonalEventPublic, UpdatePersonalEvent

router = APIRouter()


@router.get("/api/personal-events", response_model=list[PersonalEventPublic])
def get_personal_events(
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
    db: Session = Depends(get_db),
):
    return (
        db.query(PersonalEvent)
        .filter(PersonalEvent.user_id == current_user.user_id)
        .all()
    )


@router.post("/api/personal-events", response_model=PersonalEventPublic, status_code=201)
def create_personal_event(
    body: CreatePersonalEvent,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
    db: Session = Depends(get_db),
):
    ensure_profile(current_user.user_id, db)
    event = PersonalEvent(
        user_id=current_user.user_id,
        title=body.title,
        start=body.start,
        end=body.end,
        all_day=body.all_day,
        color=body.color,
    )
    db.add(event)
    db.commit()
    db.refresh(event)
    return event


@router.put("/api/personal-events/{event_id}", response_model=PersonalEventPublic)
def update_personal_event(
    event_id: str,
    body: UpdatePersonalEvent,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
    db: Session = Depends(get_db),
):
    event = (
        db.query(PersonalEvent)
        .filter(
            PersonalEvent.id == event_id,
            PersonalEvent.user_id == current_user.user_id,
        )
        .one_or_none()
    )
    if not event:
        raise HTTPException(status_code=404, detail="Personal event not found")
    event.title = body.title
    event.start = body.start
    event.end = body.end
    event.all_day = body.all_day
    event.color = body.color
    db.commit()
    db.refresh(event)
    return event


@router.delete("/api/personal-events/{event_id}")
def delete_personal_event(
    event_id: str,
    current_user: Annotated[CurrentUser, Depends(get_current_user)],
    db: Session = Depends(get_db),
):
    event = (
        db.query(PersonalEvent)
        .filter(
            PersonalEvent.id == event_id,
            PersonalEvent.user_id == current_user.user_id,
        )
        .one_or_none()
    )
    if not event:
        raise HTTPException(status_code=404, detail="Personal event not found")
    db.delete(event)
    db.commit()
    return Response(status_code=204)
